### ISSUE TITLE
Manually alter the sidebar ToC from within docs

### ISSUE THREAD

**Issue by ISSOtm** [facebook/docusaurus#6201]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Description

Docusaurus currently allows manually altering (and even outright replacing) the ToC in Docs, as per https://github.com/facebook/docusaurus/issues/3915#issuecomment-896193142, but this is not documented.

This issue is as much a question on whether this is something we can expect to rely on, as a request to document it if the answer is "yes".

### Motivation

This is useful for documentation that is generated from other sources (in this case, a `man` page): while the HTML can be injected, the ToC does not follow suit (and I wouldn't expect it to. Or is that preferable?).

### API design

Here is a doc which largely consists of externally-generated HTML, for which we additionally generate the ToC via a script.
```mdx
# rgbds(5) — object file format documentation

import generated from '!!raw-loader!@site/docs/rgbds.5.html';

<div class="manual-text" dangerouslySetInnerHTML={{ __html: generated }} />

export const toc = [
{
	"value": "NAME",
	"id": "NAME",
	"level": 2,
	"children": [
	]
},
{
	"value": "DESCRIPTION",
	"id": "DESCRIPTION",
	"level": 2,
	"children": [
	]
},
{
	"value": "FILE STRUCTURE",
	"id": "FILE_STRUCTURE",
	"level": 2,
	"children": [
{
	"value": "RPN DATA",
	"id": "RPN_DATA",
	"level": 3,
	"children": [
	]
},
	]
},
{
	"value": "SEE ALSO",
	"id": "SEE_ALSO",
	"level": 2,
	"children": [
	]
},
{
	"value": "HISTORY",
	"id": "HISTORY",
	"level": 2,
	"children": [
	]
},
];
```

### Have you tried building it?

If the existing behavior is to be accepted as official, then nothing needs to be built; otherwise, what replacement API is deemed better will need to be discussed first.

### Self-service

- [ ] I'd be willing to contribute this feature to Docusaurus myself.

**Comment by Josh-Cena** [comment]
Hi, since this feature is already in place, is it essentially a documentation request?

The TOC shape is most likely stable, but it's undocumented because up till now we haven't figured out an ergonomic way to let you use it. For example, #3915 is very likely to be solved by letting you code a TOC yourself.

**Comment by ISSOtm** [comment]
This is first a question about whether modifying the ToC like in the example is intended to be part of the API, or is just exposed internals.

If the answer is yes, then this is a documentation request. If it's no, this is a request for an alternative.

This is not intended to supersede #3915, as is explained in the comment that I originally linked to, though it can be used as a stopgap in the meantime.

**Comment by slorber** [comment]
> This is first a question about whether modifying the ToC like in the example is intended to be part of the API, or is just exposed internals.

This is internal implementation detail that is IMHO stable enough to use externally.

We could document it and make it officially a public API, but it probably requires some upfront thinking to be sure that the current API is the best way to solve your use-case

**Comment by ISSOtm** [comment]
It fits exactly my use case, since this document has the entirety of its content auto-generated.

However, if only a section were to be as such, then I'd want to be able to splice the ToC accordingly, more similarly to #3915. I think this could be arranged by having a file—`autogenned.md`—use this mechanism to manually specify the entirety of its own ToC, and then including `autogenned.md` performing the splice like requested in that PR.

Hopefully I didn't sound too confused?

**Comment by slorber** [comment]
If we add this to our doc, we should probably dogfood this on our own doc, explaining the constraints (like elements requiring a unique id for linking to work). We could showcase manual construction of a toc for generated HTML (like your case).

We should probably add this to this page (to be renamed as just "TOC"?): https://docusaurus.io/docs/markdown-features/inline-toc

Do you want to submit a PR?

I'm not sure what you mean by `autogenned.md` but I'd rather avoid showcasing the "import mdx partial" case of https://github.com/facebook/docusaurus/issues/3915, as it's not really a feature but an actual bug/limitation that we really want to solve.

**Comment by Josh-Cena** [comment]
@slorber What about we invent a "TOC enhancement" syntax ourselves and parse that in the TOC generating remark plugin?

**Comment by slorber** [comment]
@Josh-Cena I'm not sure it's really related to this issue, as here the content is in HTML and not a mdx partial import.

---

> @slorber What about we invent a "TOC enhancement" syntax ourselves and parse that in the TOC generating remark plugin?

That's something I thought about but do we really want to invent a syntax that will only serve temporarily? I'm already not a fan of inventing a non-std md syntax 😅 

The end goal is that the TOC works for imported files automatically, without asking the user to use any new fancy syntax.


---

> However, importing partials are very difficult. We either have to actually read the imported file and parse that as well

I don't think we should do that, it duplicates work to each doc importing a partial and also adds more weight to each page, as the shared toc ends up being duplicated in each doc

We'd rather favor composition and have a remark plugin that compose the toc from current doc and partials without inlining the partial tocs into the document, like your spread example, but handled automatically by the remark plugin, not hand-written by the user




> In hindsight it may be better if we have started off designing our server-side TOC structure as a flat list instead of a recursive tree

Good point, it will probably be needed to flatten that structure

So we are maybe not ready to make this "manual toc" feature an official public API 😅

**Comment by Josh-Cena** [comment]
I'm proposing this custom TOC syntax in place of handwriting the entire TOC structure which contains a lot of boilerplate. If we have that, we don't need to document the `export toc = ...` syntax at all. Also, currently it's all or nothing: you write everything or you let Remark do it all, there's no enhancement. Such an API is not ready to be documented, and we can't extend this API to allow enhancement either, we have to do it from scratch.

> The end goal is that the TOC works for imported files automatically, without asking the user to use any new fancy syntax.

As I said, it means we have to "understand" that `<Content />` is an MDX partial, not a JSX component, which isn't easy. I've tried a while back to implement a Remark plugin and call Babel to parse the JSX for me, but overall it's a very painful thing to let Remark understand JSX. It also means we have to insert an extra `{toc}` named import back into the `import Content from './_partial.mdx'` statement, and that can be intractable.

Then there's the question of (a) user wanting to hide some headings from the TOC and (b) user wanting to insert extra headings that would otherwise not be visible to Remark. We would need a way to let users handwrite & insert part of the TOC, hence the proposal for the `[[Inserted TOC]]` and `[[Imported TOC]]` syntax. Even if `[[Imported TOC]]` can be heavylifted by us, `[[Inserted TOC]]` is there to stay.

Glad I made the flattened list point through :P Going to see what we can do about it

**Comment by Josh-Cena** [comment]
This will be triaged as a feature request and we will figure out an ergonomic way to tweak the TOC structure.

**Comment by ISSOtm** [comment]
Fine by me, then. I'll keep relying on `toc` until then.

**Comment by slorber** [comment]
> As I said, it means we have to "understand" that <Content /> is an MDX partial, not a JSX component

We should look for the import statement extension, for sure it's a bit more complex but it should be achievable.

> (a) user wanting to hide some headings from the TOC 

It looks more appropriate to me to have a a syntax on the heading itself, similar to anchor links? I'm not sure how your proposal solves this use-case?

> (b) user wanting to insert extra headings that would otherwise not be visible to Remark.

For example, user using React components with some headings inside md?

That's a quite specific use-case, but still seems like a reasonable thing to solve without requiring users to write the full toc manually 

In practice, it's the use-case @ISSOtm exposed, but @ISSOtm may be satisfied by just having documentation explaining what we recommend, so it may not be so useful to implement something immediately, and doc might be good enough until it becomes more painful for a few users? 

This gives us time to think more deeply about this problem. I'd also be happy to have a way to enhance/customize the automatically generated toc object, and not sure adding proprietary markdown syntax tags is the most flexible option. I'd rather use a real function:

```md
# Title

blabla

export function toc(originalToc) {
	return [...originalToc,myExtraDocEntry];
}
```

Does it make sense?

---

> Fine by me, then. I'll keep relying on toc until then.

Yes @ISSOtm, as you see this is subject to potential breaking changes as we may flatten the toc structure 🤪  so it's not yet a good time to document but it's a good enough workaround for now.

**Comment by ISSOtm** [comment]
A bit of a tricky UX problem is splicing some headings *in the middle* of the ToC. Figuring out the correct index is less trivial than how often it'd be desirable, imo.

**Comment by Josh-Cena** [comment]
Yes, the problem with `export function toc(originalToc)` is that it's too centralized and the user can't easily customize parts of it without going to great lengths to destructure a whole TOC level. Interspersing these "artificial TOCs" sounds much more ergonomic to me.

**Comment by Josh-Cena** [comment]
Looked a bit into this. Several random thoughts:

On the point of MDX heading transclusion. The solution would be like this: `export toc = [...collectedTOC[0], ...importedTOC0, ...collectedTOC[1]]` where `collectedTOC` are all the collected headings, split according to the locations of MDX partials. `importedTOC0` is an automatically created symbol: `import Partial from './_partial.mdx'` => `import Partial, {toc as importedTOC0} from './_partial.mdx'`.

This solution is because we don't know what's actually in `_partial.mdx` unless we traverse that file as well, which is not worthwhile.

- We definitely need to flatten the TOC structure so we can freely insert slices without worrying about the hierarchy. 
  - The problem is that [selectively rendering some subtrees is actually a documented behavior](https://docusaurus.io/docs/next/markdown-features/inline-toc#custom-table-of-contents): `toc[2]` means the third top-level heading and all its subheadings. If we change to a flat list, this feature is completely lost. While we can make breaking changes, I personally think it's valuable to keep.
  - Maybe we need two data structures exported simultaneously? `toc` is a tree structure while `headings` is a flat list? The inline TOC can use the `toc` while everything else (TOC sidebar, mobile TOC) uses `headings`?
  - This still means we lose the transcluded TOC in the `toc` variable.

On the point of inserting extra anchors. This has three use-cases:

- User inserts a JSX component that contains headings. Since it's not MDX, we don't attempt to get toc from it.
- The headings are generated with HTML `<h2>`, maybe because of legacy documentation.
- User wants anchors pointing to non-heading components, e.g. `<details id="explanation">`. This is pretty uncommon, but maybe there are some documentation organized this way?

The actual syntax is open to discussion, but it would still basically be some artificial headings that will be recognized and removed by our remark plugin.

```md
<!-- this admonition-like syntax encapsulates some artificial headings
that will be present in the TOC but removed from the content -->
:::toc

## Explanation {#explanation}

:::
```

Something like that...

This part doesn't require any refactors, because the remark plugin will see artificial headings the same as normal ones.

On the point of hiding headings away from the TOC. This is the tricky part. Note that `{#anchor}` is actually a widely-adopted Markdown feature, so up to this point Docusaurus hasn't done anything unique yet. Maybe we can have another syntax that's very similar to this anchor syntax?

```md
## Hidden heading {!#heading}
```

This will still allow us to set the anchor ID, but the `!` tells Docusaurus that this heading shouldn't be indexed.

In conclusion, we will:

- Solve the MDX TOC transclusion under the hood, without inventing any new API. This would mean, however, a change to our TOC structure.
- Allow inserting extra headings with a special "container", either `:::toc` or `[[toc]]` where the contents are indexed by the remark plugin and then removed.
- Allow hiding headings with a special anchor ID syntax.

**Comment by Josh-Cena** [comment]
The artificially inserted TOC will likely be useful for our API doc: https://docusaurus.io/docs/next/api/plugins/@docusaurus/plugin-content-docs

I envision something similar to Yarn's doc: https://yarnpkg.com/cli/workspaces/foreach#options where every line in the table can have its toc link

**Comment by slorber** [comment]
Yarn table is using h3

Couldn't we also use this?

**Comment by Josh-Cena** [comment]
I'm not sure if it's good practice, and MDX refuses to render headings within table rows (for good reason):

![image](https://user-images.githubusercontent.com/55398995/151360512-a43a5f7d-cd71-4a59-8946-f98e80173749.png)

If we use `<h3>` then we lose the toc function anyways.

**Comment by jasikpark** [comment]
would like to use a component for my headings, while including them in the TOC.

currently approach is to change `<ChangelogHeading version="0.1.8" date="2022-12-20"/>` to 

```mdx
<ChangelogHeading date="2022-12-20">

## 0.1.8

</ChangelogHeading>
```

so that Docusaurus picks up the MDX heading to put in the TOC while still rendering as

```jsx
<hgroup style={{ display: 'flex', flexWrap: 'wrap', alignItems: 'baseline', gap: '2em' }}>
      <h2>0.1.8</h2>
      <time dateTime="2022-12-20">2022-12-20</time>
</hgroup>
```

**Comment by jasikpark** [comment]
interesting! thanks for the suggestions, i think i'll just stick with my current solution in that case 👍

**Comment by shmuelisrl** [comment]
I was trying to add headings that show up in the TOC, but I've encountered an issue.
when you add a Heading in a bullet point it doesn't show up in the toc and i want the option to have it show up.
I'm actually just trying to indent certain headings, and I'm doing this
```mdx
<div class="nobullet">

# Inputs
* ## A
  * details
* ## B
  * details
* ## C
  * details
* ## D
  * details

</div>
```
I'd like to have the ability to set any word to be of not be in the TOC no matter where it is.
like this:
```md
word{is-in-TOC}(custom id)

* word{is-in-TOC}(custom id 2)

or
## word{is-not-in-TOC}
```

**Comment by foot** [comment]
Trying to automate the `export const toc` trick mentioned in this issue description, I very naively hacked together

```js
// Use cheerio as docusaurus uses it
import * as cheerio from "cheerio";

// Read in raw html string and generate docusaurus toc format
export const genToc = function (apiRaw) {
  const $ = cheerio.load(apiRaw);
  return $("h2, h3, h4, h5, h6")
    .toArray()
    .map((header) => {
      const $header = $(header);

      const level = parseInt($header[0].name[1], 10);
      const value = $header.text();
      const id = $header.attr("id");
      return {
        level,
        value,
        id,
      };
    });
};
```

So one can then

```jsx
import GeneratedAPI from './_api.mdx';

This is the generated api

<GeneratedAPI />

import { genToc } from '../_components/api-toc.js';
import apiRaw from "!!raw-loader!./_api.mdx";
export const toc = genToc(apiRaw);
```

However this approach _seems_ to add 300kb to the page, which I would guess is mostly cheerio coming down the line. (_api.mdx is only 20kb)

I'm unfamiliar with SSR but could this be forced to be done at "SSG" time during a `yarn build` rather than in the browser? Would changing the above to use browser API's instead of cheerio still build okay?

**Comment by slorber** [comment]
@foot yes the TOC has to be computed server-side otherwise it would be first invisible and then "pop" once React hydrates. Here you include cherrio but also the markdown file as a source string (so each doc is now both a React component + a string).

This really should be done at build time, ideally as a remark plugin.

**Comment by nullromo** [comment]
Adding some more detail to my previous comment now that I learned a little more about Remark/Rehype plugins.

By dumping out the `mainNode` object from [the rehype-toc plugin](https://github.com/JS-DevTools/rehype-toc/blob/master/src/rehype-toc.ts#L19), I found that the structure of the document looks weird to me. For normal headings, I get nodes that look like 
```json
{
    "type": "element",
    "tagName": "h2",
    "properties": { "id": "asdf" },
    "children": [
        {
            "type": "text",
            "value": "ASDF",
            "position": {
                "start": { "line": 111, "column": 4, "offset": 2968 },
                "end": { "line": 111, "column": 12, "offset": 2976 }
            }
        }
    ],
    "position": {
        "start": { "line": 111, "column": 1, "offset": 2965 },
        "end": { "line": 111, "column": 12, "offset": 2976 }
    }
},
```
but the component ones look like some kind of embedded JSX instead, so the ToC plugin is unable to detect the header elements within them
```json
{
    "type": "jsx",
    "value": "<Paragraph\n    aContent={\n        <>\n            <hr />\n            <h2>ASDF 2</h2>\n            <EndpointTemplate\n                description= ........", // truncated for brevity
    "position": {
        "start": { "line": 124, "column": 1, "offset": 3280 },
        "end": { "line": 197, "column": 3, "offset": 5500 },
        "indent": [
            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1
        ]
    }
},
```

As you can see, once we get to a JSX component, it stops being a tree and just flattens everything into 1 node.

Somehow somewhere this JSX is expanded into an HTML tree, but I don't know how or where. Is there any way to run some kind of plugin or transformation **after** the JSX is expanded?

**Comment by slorber** [comment]
@nullromo we are not using remark-toc, if it does not behave the way you want we can't do anything for you.

---

In any case, you can provide an explicit toc structure yourself on any doc that overrides the one we compute:

```tsx
export const toc = [
  {
    value: "Label1",
    id: "anchor1",
    level: 2,
    children: [],
  },
  {
    value: "Label2",
    id: "anchor2",
    level: 2,
    children: [
      {
        value: "Label3",
        id: "anchor3",
        level: 3,
        children: [],
      },
    ],
  },
];
```

You can also take over how this structure is rendered with swizzle. 

I agree that's probably not ideal and requires maintenance, but that works

---

If your goal is to only have placeholders in headings (and not dynamic headings that are present/absent conditionally), that can probably be implemented with a remark plugin:

```md
## Getting Started with %%PRODUCT_NAME%%
```

We don't provide such plugins but you'll find useful information here and code examples here:
https://github.com/facebook/docusaurus/issues/395

This code example is probably the best to get started: https://github.com/facebook/docusaurus/issues/395#issuecomment-1496701245
You can probably tweak it a bit so that it also substitutes variables in markdown headings,

**Comment by nullromo** [comment]
@slorber Thanks for the advice. I was able to come up with a plugin just like how you described. Unfortunately, I still didn't find a way to actually parse through the JSX nodes within the AST, so I just have to use a regex to determine which nodes I want to remove. I'll post the code here (with names changed) for reference.

### Plugin
```ts
const visit = require('unist-util-visit');

// plugin that removes certain components
const myPlugin = () => {
    // determine which nodes to filter out based on the product type
    const nodeRegex = (() => {
        switch (process.env.PRODUCT_TYPE) {
            case 'PRODUCT_A':
                return /(ProductBParagraph|ProductCParagraph)/;
            case 'PRODUCT_B':
                return /(ProductAParagraph|ProductCParagraph)/;
            case 'PRODUCT_C':
                return /(ProductAParagraph|ProductBParagraph)/;
            default:
                return /^$/;
        }
    })();
    return async (
        /** @type {import("unist").Node<import("unist").Data>} */ ast,
    ) => {
        // this variable will become true when we hit the opening tag for the
        // node to be removed, and it will become false when we hit the next
        // tag. This means that it will be true for all the nodes between the
        // opening and closing tags
        let removing = false;

        // traverse the tree
        visit(ast, {}, (child, index, parent) => {
            // remember if we removed the current node or not
            let removed = false;

            // removes the current node if we are in removing mode
            const removeNodeIfNeeded = () => {
                // check if in removing mode
                if (removing) {
                    // remove the node
                    //console.log('removing', child);
                    parent.children.splice(index, 1);
                    // remember that we removed the node
                    removed = true;
                }
            };

            // remove the current node if necessary
            removeNodeIfNeeded();

            // if the node is a JSX node, see if we hit the opening/closing tag
            if (child.type === 'jsx') {
                // if the value matches, then it's the right tag
                if (
                    // @ts-ignore
                    nodeRegex.test(child.value)
                ) {
                    // toggle removing mode
                    removing = !removing;
                    // if we just toggled on, remove this node
                    removeNodeIfNeeded();
                }
            }

            // if we removed the node, return SKIP, otherwise just return
            if (removed) {
                return [visit.SKIP, index];
            }
            return;
        });
    };
};
```

### Components
```jsx
export enum ProductType {
    PRODUCT_A = 'PRODUCT_A',
    PRODUCT_B = 'PRODUCT_B',
    PRODUCT_C = 'PRODUCT_C',
}

const useProductType = () => {
    return useDocusaurusContext().siteConfig.customFields
        .productType as ProductType;
};

const makeProductTypeParagraph = (
    productFilter: ProductType,
) => {
    return (props: React.PropsWithChildren) => {
        const productType = useProductType();
        if (productType === productTypeFilter) {
            return <>{props.children}</>;
        }
        return null;
    };
};

export const ProductAParagraph = makeProductTypeParagraph(ProductType.PRODUCT_A);
export const ProductBParagraph = makeProductTypeParagraph(ProductType.PRODUCT_B);
export const ProductCParagraph = makeProductTypeParagraph(ProductType.PRODUCT_C);
```

### Markdown
```mdx
<ProductAParagraph>

---

## Product A Heading 1

<Thing
    cool='yeah'
/>

---

## Product A Heading 2

<>
    <MyComponent
        nice='wow'
    />
</>

</ProductAParagraph>

<ProductBParagraph>

---

## Product B Heading 1

awesome

</ProductBParagraph>

```

I don't necessarily like the regex matching technique here because it kind of blindly removes stuff without _really_ knowing what's going on. For example, if my approach is to remove everything between the opening and closing tags (`<ProductAParagraph>` and `</ProductAParagraph>`), then what happens when I try to use a self-closing tag (`<ProductAParagraph />`)? Things are going to get messed up.

So if there's any way that you know of to actually process JSX from inside a remark plugin, I'd love to hear about it.

In any case, thanks a lot for the help! Glad I now have a suitable workaround 🎉

**Comment by slorber** [comment]
> So if there's any way that you know of to actually process JSX from inside a remark plugin, I'd love to hear about it.

First: I'd recommend implementing this in Docusaurus v3 (MDX is now at v2), currently in alpha.

That was just to give you a direction, I have not implemented this myself. If you want to build this properly you have to learn MDX / Unified and how all things work together: you can't skip reading the doc and investing some time.

You should inspect the produced AST tree and see which nodes you want to remove. Most likely the nodes will look like this:

```tsx
{
      "type": "mdxJsxFlowElement",
      "name": "div",
      "attributes": [],
      "children": [/* content */]
}
```

What I would do is use a visitor to visit all `mdxJsxFlowElement`, and remove some of them conditionally. You don't need a regexp for that and self-closing tags are handled: there will be no children in such nodes.

Again this is just an idea and direction: you'll have to figure out the details yourself and learn how these things work.

**Comment by jeluard** [comment]
Exporting the global `toc` apparently doesn't work anymore with docusaurus v3. Is there an alternative to programmatically set the toc?

**Comment by slorber** [comment]
@jeluard while improving the TOC to support imported docs, I noticed a strange behavior. 

Related discussion:
https://github.com/facebook/docusaurus/pull/7530#discussion_r1458087876


Apparently, exporting `toc` would only work on docs that don't contain any heading.

![CleanShot 2024-03-21 at 13 13 52](https://github.com/facebook/docusaurus/assets/749374/75679bb5-a77b-4b81-815f-bc2ab2086c3c)

But as soon as you have headings (>= level 2), they get used in priority over your exported TOC.

![CleanShot 2024-03-21 at 13 14 04](https://github.com/facebook/docusaurus/assets/749374/8ca7b98d-ddbf-407d-8f28-7e3352b380a1)

---

This didn't look good to me, so I fixed this behavior for Docusaurus v3.2/canary to always let you the ability to override the generated toc:

https://stackblitz.com/edit/github-zjz2fr?file=docs%2Fintro.mdx,package.json

![CleanShot 2024-03-21 at 13 18 38](https://github.com/facebook/docusaurus/assets/749374/950ca00b-f214-4352-9207-083eebeadbe9)

**Comment by jeluard** [comment]
@slorber I can confirm that it works as expected with canary docusaurus. Thanks!

**Comment by rtrbt** [comment]
Just in case someone else reads this: Some of the custom ToCs in this issue set their children like this:

```mdx
export const toc = [{
        value: "Label1",
        id: "anchor1",
        level: 2,
        children: [],
    },{
        value: "Label2",
        id: "anchor2",
        level: 2,
        children: [{
            value: "Label3",
            id: "anchor3",
            level: 3,
            children: [],
        }],
    }
]
```

to create the same ToC as

```markdown
## Label1 {#anchor1}

## Label2 {#anchor2}

### Label3 {#anchor3}
```

However (probably since https://github.com/facebook/docusaurus/pull/6729 ) setting children is not necessary anymore and the child entries don't show up in the ToC. This works:

```mdx
export const toc = [{
        value: "Label1",
        id: "anchor1",
        level: 2,
    },{
        value: "Label2",
        id: "anchor2",
        level: 2,
    },{
        value: "Label3",
        id: "anchor3",
        level: 3,
    }
]
```

**Comment by slorber** [comment]
Yes I confirm in 2021 we had a nested structure, and now there's no `children` anymore, the TOC structure is flat.

**Comment by ISSOtm** [comment]
Thank you for the notice! I [updated upstream](//github.com/gbdev/rgbds-www/commit/84d09523d8a0b2be62d4ec4afdbb42edf31be068), and edited the OP accordingly. This works much better :)

**Comment by slorber** [comment]
@axmmisaka if you want to put headings inside tabs and expect the TOC to update according to the selected tab, there's another issue for that: https://github.com/facebook/docusaurus/issues/5343

**Comment by axmmisaka** [comment]
> @axmmisaka if you want to put headings inside tabs and expect the TOC to update according to the selected tab, there's another issue for that: #5343

Thanks for the reply.
I would assume that this issue has been designated as a wontfix and contributing to upstream would be the way to have it. 
My understanding is that part of the difficulty arise from the fact that we will need to tell remark to parse in a way which records whether or not the header is in a `Tab`.
A workaround I could think of is to allow the author to supply a header and switch between them on-the-go. This would look ugly but would bypass the remark issue.

**Comment by gdude2002** [comment]
Since the Docusaurus docs [mentioned leaving use-cases on this issue](https://docusaurus.io/docs/markdown-features/toc#customizing-table-of-contents-generation), I figured I'd throw my (fairly niche) hat into the ring.

On [this page](https://docs-beta.kordex.dev/bots/commands/converters/bundled), I have some fairly custom React components, representing classes that convert strings to rich types:

![Image](https://github.com/user-attachments/assets/809f8f68-5db8-41b9-837b-3bdf55988c12)

Of course, the current TOC approach gives me this: 

![Image](https://github.com/user-attachments/assets/b6dae942-c882-4740-be6d-95125dc3f484)

In this case, it would be nice (though not enough of a need for me to go generating my own full TOCs) to be able to list these components as children under their headers. Currently, I don't think I have any serious way to do this, and I know it's fairly niche, but it _is_ a use-case!
